common/src/main/java/com/spredfast/kafka/connect/s3/GroupedByTopicLayout.java

```diff
@@ -0,0 +1,71 @@
+package com.spredfast.kafka.connect.s3;
+
+import org.apache.kafka.common.TopicPartition;
+
+import java.util.function.Supplier;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class GroupedByTopicLayout implements Layout {
+
+    private final Supplier<String> dateSupplier;
+
+    public GroupedByTopicLayout(Supplier<String> dateSupplier) {
+        this.dateSupplier = dateSupplier;
+    }
+
+    public Layout.Builder getBuilder() {
+        return new Builder(dateSupplier);
+    }
+
+    public Layout.Parser getParser() {
+        return new Parser();
+    }
+
+    static class Builder implements Layout.Builder {
+
+        private final Supplier<String> dateSupplier;
+
+        public Builder(Supplier<String> dateSupplier) {
+            this.dateSupplier = dateSupplier;
+        }
+
+        @Override
+        public String buildBlockPath(BlockMetadata blockMetadata) {
+            final TopicPartition tp = blockMetadata.getTopicPartition();
+            return String.format("%s/%s/%05d-%012d", tp.topic(), dateSupplier.get(), tp.partition(),
+                    blockMetadata.getStartOffset());
+        }
+
+        @Override
+        public String buildIndexPath(TopicPartition topicPartition) {
+            return String.format("%s/last_chunk_index.%05d.txt", topicPartition.topic(), topicPartition.partition());
+        }
+    }
+
+    public static class Parser implements Layout.Parser {
+
+        private static final Pattern KEY_PATTERN = Pattern.compile(
+                // match the / or the start of the key, so we shouldn't have to worry about the prefix
+                "(/|^)"
+                        // assuming no / in topic names
+                        + "(?<topic>[^/]+?)/"
+                        + "(?<date>[^/]+?)/"
```

Comment:
It's clear why this regex contains `date` group. But I don't understand why the same regex from  `GroupedByDateLayout` doesn't contain it.

Reply:
Because the date is ignored in both cases (it's only used for ordering objects in the listing).

In `GroupedByDateLayout`, it's ignored together with the prefix. In this one, it's between the topic and partition, so it's also ignored but needs to be mentioned. It doesn't have to be named and doesn't need to capture the matching sub-string though.